### PR TITLE
fix(titus): Handle/report incomplete target group cache data

### PR DESCRIPTION
**Pull request by robzienert**
The NPE was caused by traversing a map-of-maps without actually being certain the data is there. Seems like the actual usage here is reasonable, but the cache data was corrupted / inconsistent in Redis. In this case, we'll just not report health data.

I took this as an opportunity to clean up the class a little as well, but the changes are in separate commits.

### CODE REVIEW

**Comment by jrsquared** [comment]
🥈

File: clouddriver-titus/src/main/groovy/com/netflix/spinnaker/clouddriver/titus/caching/providers/TitusTargetGroupServerGroupProvider.java [hunk 1]
```diff
@@ -133,4 +125,75 @@ public TitusTargetGroupServerGroupProvider(Cache cacheView) {
...
+
+  private static Map<String, Object> getTargetGroupHealth(String instanceKey,
+                                                   Map<String, String> targetGroupDetails,
+                                                   CacheData healthData) {
+    try {
+      if (healthDataContainsTargetGroups(healthData)) {
+        Map targetGroupHealth = getTargetGroupHealthData(targetGroupDetails, healthData);
+
+        if (!targetGroupHealth.isEmpty()) {
+          Map<String, Object> health = new HashMap<>();
```
Comment (robzienert) [inline clouddriver-titus/src/main/groovy/com/netflix/spinnaker/clouddriver/titus/caching/providers/TitusTargetGroupServerGroupProvider.java]: Also a small bit of efficiency here. Doing a static initialization of a hashmap `e.g. new HashMap<>() { /* put commands */ };` creates a temporary class and doesn't buy us much terseness anyhow.

File: clouddriver-titus/src/main/groovy/com/netflix/spinnaker/clouddriver/titus/caching/providers/TitusTargetGroupServerGroupProvider.java [hunk 2]
```diff
@@ -48,77 +49,68 @@ public TitusTargetGroupServerGroupProvider(Cache cacheView) {
...
-                  }
-                }
-              } catch (Exception e) {
-                log.error("failed to load health for " + instanceKey, e);
+
+              Optional<LoadBalancerInstance> instance = getInstanceHealth(instanceKey, instanceDetails, targetGroupDetails);
+              if (instance.isPresent()) {
+                targetGroupInstances.add(instance.get());
+              } else {
+                log.error(format(
```
Comment (ajordens) [inline clouddriver-titus/src/main/groovy/com/netflix/spinnaker/clouddriver/titus/caching/providers/TitusTargetGroupServerGroupProvider.java]: don't really need `String.format()` ... `"Detected potentially .. (targetGroup: {}, serverGroup: {})"` should suffice?

File: clouddriver-titus/src/main/groovy/com/netflix/spinnaker/clouddriver/titus/caching/providers/TitusTargetGroupServerGroupProvider.java [hunk 3]
```diff
@@ -133,4 +125,75 @@ public TitusTargetGroupServerGroupProvider(Cache cacheView) {
...
+    } catch (NullPointerException e) {
+      return Optional.empty();
+    }
+
+    CacheData healthData = cacheView.get(HEALTH.ns, healthKey);
+
+    Map<String, Object> health = getTargetGroupHealth(instanceKey, targetGroupDetails, healthData);
+
+    return Optional.of(new LoadBalancerInstance(
+      ((Map) instanceDetails.get("task")).get("id").toString(),
```
Comment (ajordens) [inline clouddriver-titus/src/main/groovy/com/netflix/spinnaker/clouddriver/titus/caching/providers/TitusTargetGroupServerGroupProvider.java]: I guess it's safe to say an instance will always have a task will always have an id ... we weren't guarding against this previously and I can't think of how this wouldn't hold true

File: clouddriver-titus/src/main/groovy/com/netflix/spinnaker/clouddriver/titus/caching/providers/TitusTargetGroupServerGroupProvider.java (hunk 2 above)
Comment (ajordens) [reply clouddriver-titus/src/main/groovy/com/netflix/spinnaker/clouddriver/titus/caching/providers/TitusTargetGroupServerGroupProvider.java]: thinking out loud --- I wonder where else we could start looking for a signal that caches are potentially inconsistent in some way and emit a metric (short of a user telling us)

Comment (robzienert) [reply clouddriver-titus/src/main/groovy/com/netflix/spinnaker/clouddriver/titus/caching/providers/TitusTargetGroupServerGroupProvider.java]: This is a potential problem anywhere we're resolving relationship data between two types, right? If that assertion is correct, I'd tend to think the problem could pop up in quite a number of places, like in EC2 load balancer health checks.

Comment (ajordens) [reply clouddriver-titus/src/main/groovy/com/netflix/spinnaker/clouddriver/titus/caching/providers/TitusTargetGroupServerGroupProvider.java]: Yes, I believe that's true.

Something to keep on the back burner, it'd be nice to more readily detect when things are consistently inconsistent.

**Pull request merged**